### PR TITLE
Using same refolding strategy for named cofixpoints in tactic "simpl" as is done for named fixpoints

### PR DESCRIPTION
**Pull request by herbelin**
We mostly reuse for named cofixpoints the code existing to refold named fixpoints.

Fixes #4056 and makes `simpl`'s behavior more uniform and a bit closer to `cbn`.

The 5 first commit are code restructuration (no change of semantics in principle). The 6th is a shortcut in the code. Only the last commit changes the behavior wrt cofixpoints refolding.

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

Note: I'm considering eventually introducing refolding also in `lazy` and `cbv`.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by herbelin** [comment]
CI green at this stage (except coq_library_undecidability and http as in other PRs).

File: pretyping/reductionops.mli [hunk 1]
```diff
@@ -58,7 +58,8 @@ module Stack : sig
 
   val pr_app_node : (EConstr.t -> Pp.t) -> app_node -> Pp.t
 
-  type case_stk
+  type case_stk =
+    case_info * EInstance.t * EConstr.t array * EConstr.t pcase_return * EConstr.t pcase_invert * EConstr.t pcase_branch array
```
Comment (ppedrot) [inline pretyping/reductionops.mli:62]: It seems you only build values of this type. If that's the case (no pun intended) don't expose the type definition but a builder function instead.

Comment (herbelin) [reply pretyping/reductionops.mli:62]: It is the case... Done

Comment (ppedrot) [reply pretyping/reductionops.mli:62]: You didn't hide the type though.

Comment (herbelin) [reply pretyping/reductionops.mli]: Lol, looks like I was a bit absent-minded. Done

**Comment by ppedrot** [comment]
@coqbot run full ci

File: pretyping/tacred.ml [hunk 2]
```diff
@@ -322,64 +341,125 @@ let deactivate_delta allowed_reds =
   (* Act both on Delta and transparent state as not all reduction functions work the same *)
   RedFlags.(red_add_transparent (red_sub allowed_reds fDELTA) TransparentState.empty)
 
-(* [compute_consteval] stepwise expands an arbitrary long sequence of
+(* [compute_constant_elimination] stepwise expands an arbitrary long sequence of
     reversible constants, eventually refolding to the initial constant
     for unary fixpoints and to the last constant encapsulating the Fix
     for mutual fixpoints *)
 
-let compute_consteval ((cache,_),allowed_reds as cache_reds) env sigma ref u =
```
Comment (ppedrot) [inline pretyping/tacred.ml:330]: I think we should introduce a proper record for reduction caches.

**Comment by ppedrot** [comment]
@coqbot merge now

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

File: pretyping/tacred.ml (hunk 2 above)
Comment (herbelin) [reply pretyping/tacred.ml:330]: Eventually yes. I could do it as soon as this PR if you insist, or, maybe, simpler to me, at the time of rebasing PR #18580 which also adds a new component.

**Pull request merged**